vendor buildkit 6861f17f15364de0fe1fd1e6e8da07598a485123

tiborvass opened the pull request:
Signed-off-by: Tibor Vass <tibor@docker.com>

thaJeztah commented:
Failures look related;

```
2020-10-28T17:53:45.190Z] === Errors
[2020-10-28T17:53:45.190Z] builder/dockerfile/dispatchers_test.go:142:37: unknown field 'KeyValuePairOptional' in struct literal of type instructions.ArgCommand
[2020-10-28T17:53:45.190Z] builder/dockerfile/dispatchers_test.go:398:34: unknown field 'KeyValuePairOptional' in struct literal of type instructions.ArgCommand

thaJeztah commented:
Created a list of diffs per dependency:

- https://github.com/moby/buildkit/compare/4d1f260e8490ec438ab66e08bb105577aca0ce06...212a0b14394dd09e9ec5e112469c526a14a93b2a
- https://github.com/tonistiigi/fsutil/compare/ae3a8d753069d0f76fbee396457e8b6cfd7cb8c3...c3ed55f3b48161fd3dc42c17ba09e12ac52d57dc
- https://github.com/opentracing/opentracing-go/compare/1361b9cd60be79c4c3a7fa9841b3c132e40066a7...v1.2.0
- https://github.com/google/shlex/compare/e7afc7fbc51079733e9468cdfd1efcd7d196cd1d...6f45313302b9c56850fc17f99e40caebce98c716
    - this is downgrading to an older version: https://github.com/google/shlex/compare/6f45313302b9c56850fc17f99e40caebce98c716...e7afc7fbc51079733e9468cdfd1efcd7d196cd1d
- https://github.com/opentracing-contrib/go-stdlib/compare/b1a47cfbdd7543e70e9ef3e73d0802ad306cc1cc...v1.0.0
- https://github.com/mitchellh/hashstructure/compare/2bca23e0e452137f789efbc8610126fd8b94f73b...v1.0.0
- https://github.com/containerd/containerd/compare/v1.4.1...d4e78200d6da62480c85bf6f26b7221ea938f396
    - raw diff (as it's different branches); https://github.com/containerd/containerd/compare/v1.4.1..d4e78200d6da62480c85bf6f26b7221ea938f396
    - I _think_ the relevant commits we need for buildkit are https://github.com/containerd/containerd/pull/4445; how realistic would it be to get those backported to the v1.4 branch?
- https://github.com/containerd/go-runc/compare/7016d3ce2328dd2cb1192b2076ebd565c4e8df0c...16b287bc67d069a60fa48db15f330b790b74365b
- https://github.com/containerd/stargz-snapshotter (new dependency; Apache2 License, so should be ok)

thaJeztah commented on vendor.conf:
```diff
@@ -27,13 +27,13 @@ github.com/imdario/mergo                            1afb36080aec31e0d1528973ebe6
-github.com/opentracing-contrib/go-stdlib            b1a47cfbdd7543e70e9ef3e73d0802ad306cc1cc
-github.com/mitchellh/hashstructure                  2bca23e0e452137f789efbc8610126fd8b94f73b
+github.com/opentracing/opentracing-go               d34af3eaa63c4d08ab54863a4bdd0daa45212e12 # v1.2.0
+github.com/google/shlex                             6f45313302b9c56850fc17f99e40caebce98c716
```
this is downgrading to an older version; could you update the version in buildkit instead?

thaJeztah commented:
I know containerd doesn't have a stable Go API (yet); still trying to see what chances we have to stay on release branches / released versions, so that we can have bug- and security fixes backported upstream, and to make our dependency graph more `go modules` friendly; I think the relevant commits we need for buildkit are containerd/containerd#4445; how realistic would it be to get that backported to the v1.4 branch?

tiborvass commented:
Sorry this is not ready, I need to fix things still.

AkihiroSuda commented:
> https://github.com/containerd/stargz-snapshotter (new dependency; Apache2 License, so should be ok)

@ktock Could we reorganize the package structure not to require this dependency? (Not a blocker for merging this PR)

thaJeztah commented:
> @ktock Could we reorganize the package structure not to require this dependency? (Not a blocker for merging this PR)

👍 yes, if we don't need it in this code, that would be better of course!

thaJeztah commented:
Some upstream buildit PRs were opened;

- [ ] https://github.com/moby/buildkit/pull/1762 was opened to get rid of the stargz-snapshotter dependency
- [ ] https://github.com/moby/buildkit/pull/1767 to update google/shlex in buildkit to the same version as is already used in this repo
- [ ] https://github.com/moby/buildkit/pull/1766 to move the dockerignore package to buildkit; we should deprecate the package in this repository (remove the code, and replace with aliases with a `Deprecated:` comment, so that the buildkit is the canonical place for that

Commit:
vendor buildkit 6861f17f15364de0fe1fd1e6e8da07598a485123

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
builder-next: Refactor using buildkit's resolver pool

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
builder-next: simple refactor

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
builder-next: refactor to use flighcontrol and remove resolveOnce

Signed-off-by: Tibor Vass <tibor@docker.com>

tiborvass commented:
@thaJeztah @tonistiigi updated. I will open a separate PR for sync-ing vendoring hashes.

thaJeztah commented on builder/dockerignore/dockerignore_test.go:
```diff
@@ -1,69 +0,0 @@
-package dockerignore // import "github.com/docker/docker/builder/dockerignore"
```
Can you combine the "use buildkit dockerignore" with the vendor commit?

Also should add a "deprecated.go" that deprecated it, but aliases it to to the new location (to help the transition); there's quite some consumers of it; https://grep.app/search?current=2&q=github.com/docker/docker/builder/dockerignore

tiborvass replied:
This is not `pkg/` but fine

thaJeztah replied:
Does it make a difference? afaik the `pkg/` prefix was a _convention_, but now discouraged and deprecated

thaJeztah commented:
This failure looks related;

```
=== RUN   TestDockerSuite/TestBuildContChar
    --- FAIL: TestDockerSuite/TestBuildContChar (2.41s)
        docker_cli_build_test.go:5627: assertion failed: expression is false: strings.Contains(result.Combined(), "Step 2/2 : RUN echo hi \\\n")
```

I suspect my changes in https://github.com/moby/buildkit/pull/1559 broke that test. Looks like it was added in https://github.com/moby/moby/commit/eaf0b5708fc91ab0c78eeee1a2c9c1251439d4a8 (https://github.com/moby/moby/pull/27873), but AFAICS that was testing a bare (non-escaped) `\` at the end of the line; https://github.com/moby/moby/issues/27545

Commit:
builder-next: fix --cache-from panic

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
integration-cli: remove TestBuildContChar

This old test is failing after an edge-case change in dockerfile
parsing considered a bugfix: https://github.com/moby/buildkit/pull/1559

Instead of fixing the test, I suggest removing it as there are already
tests for it in BuildKit.

Signed-off-by: Tibor Vass <tibor@docker.com>

tiborvass commented:
@thaJeztah @tonistiigi it's updated and the s390x failure seems unrelated. `--cache-from` works for both inline cache local, inline-cache remote, and with remote cache exported with mode=max.

thaJeztah commented:
Thanks!

> On 14 Nov 2020, at 07:32, Tibor Vass <notifications@github.com> wrote:
> 
> ﻿
> @thaJeztah @tonistiigi it's updated and the s390x failure seems unrelated. --cache-from works for both inline cache local, inline-cache remote, and with remote cache exported with mode=max.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

tonistiigi commented:
Building buildkit:master with this PR:
```
root@81f0e16cc830:/tmp/buildkit# docker build .
[+] Building 7.0s (15/30)
 => [internal] load build definition from Dockerfile                                                                                                    0.0s
 => => transferring dockerfile: 13.03kB                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                       0.0s
 => => transferring context: 56B                                                                                                                        0.0s
 => resolve image config for docker.io/docker/dockerfile:1.1-experimental                                                                               1.9s
 => docker-image://docker.io/docker/dockerfile:1.1-experimental@sha256:de85b2f3a3e8a2f7fe48e8e84a65f6fdd5cd5183afa6412fff9caa6871649c44                 0.8s
 => => resolve docker.io/docker/dockerfile:1.1-experimental@sha256:de85b2f3a3e8a2f7fe48e8e84a65f6fdd5cd5183afa6412fff9caa6871649c44                     0.0s
 => => sha256:de85b2f3a3e8a2f7fe48e8e84a65f6fdd5cd5183afa6412fff9caa6871649c44 1.69kB / 1.69kB                                                          0.0s
 => => sha256:8c69d118cfcd040a222bea7f7d57c6156faa938cb61b47657cd65343babc3664 521B / 521B                                                              0.0s
 => => sha256:08fb2b2ca3d58e19d791e73dea16126df37608115532e748d63525688df457a5 897B / 897B                                                              0.0s
 => => sha256:61261561661960014533790f5d6c42f5b88362db7e005f35248c20f075900a61 8.88MB / 8.88MB                                                          0.5s
 => => extracting sha256:61261561661960014533790f5d6c42f5b88362db7e005f35248c20f075900a61                                                               0.2s
 => [internal] load metadata for docker.io/library/alpine:3.12                                                                                          1.3s
 => [internal] load metadata for docker.io/tonistiigi/xx:golang@sha256:6f7d999551dd471b58f70716754290495690efa8421e0a1fcf18eb11d0c0a537                 1.5s
 => [internal] load metadata for docker.io/library/golang:1.13-buster                                                                                   1.5s
 => CANCELED [internal] load build context                                                                                                              1.0s
 => => transferring context: 38.34MB                                                                                                                    1.0s
 => [alpine 1/1] FROM docker.io/library/alpine:3.12@sha256:c0e9560cda118f9ec63ddefb4a173a2b2a0347082d7dff7dc14272e7841a5b5a                             1.6s
 => [gobuild-minimal 1/3] FROM docker.io/library/golang:1.13-buster@sha256:66a3f6817c129f48c9cc9b96816a1c0b6dcb399393196e3581d9367b55ab29f2             1.6s
 => CANCELED FROM docker.io/tonistiigi/binfmt:buildkit                                                                                                  1.0s
 => => resolve docker.io/tonistiigi/binfmt:buildkit                                                                                                     1.6s
 => => sha256:c43f2553830c3fe4846f88e6ea796237c83ebda9198896312bb1154795965696 7.78MB / 7.78MB                                                          0.6s
 => => sha256:9846723457345f5bc7247fd9681b51868c5b635ea77036d282ab774d2fbc0893 2.63MB / 2.63MB                                                          0.5s
 => => sha256:9da471824dab89882c995081b262df183c0d4522d4cef3656c28e54883c04df1 2.00kB / 2.00kB                                                          0.0s
 => => sha256:eee96be38e5e33c36b4c70128eeb781adc144ea342dc0dabc61ea3234416ec41 739B / 739B                                                              0.0s
 => => sha256:ceae2fe9603c9985f8874fda667cbea184ce0837ba156ad53873668ad5c9e81e 6.06kB / 6.06kB                                                          0.0s
 => => extracting sha256:c43f2553830c3fe4846f88e6ea796237c83ebda9198896312bb1154795965696                                                               0.4s
 => [xgo 1/1] FROM docker.io/tonistiigi/xx:golang@sha256:6f7d999551dd471b58f70716754290495690efa8421e0a1fcf18eb11d0c0a537                               0.3s
 => => resolve docker.io/tonistiigi/xx:golang@sha256:6f7d999551dd471b58f70716754290495690efa8421e0a1fcf18eb11d0c0a537                                   0.0s
 => => sha256:6f7d999551dd471b58f70716754290495690efa8421e0a1fcf18eb11d0c0a537 3.29kB / 3.29kB                                                          0.0s
 => => sha256:3265075f008d7c003eea3e5c4da277cd5049c0aa31d39d1ec35272168f8bf9dd 523B / 523B                                                              0.0s
 => => sha256:ebffcbb79133382ea62829f690a0273304a3f4d83cd4876b58af645356512348 466B / 466B                                                              0.0s
 => => sha256:a104d17248520be6cb06ad88875c391c27c5e9e2a7f474d8ad6222d6bbe02445 616B / 616B                                                              0.1s
 => => extracting sha256:a104d17248520be6cb06ad88875c391c27c5e9e2a7f474d8ad6222d6bbe02445                                                               0.0s
 => ERROR [git 1/1] RUN apk add --no-cache git xz                                                                                                       0.7s
 => [gobuild-minimal 2/3] COPY --from=xgo / /                                                                                                           0.1s
 => ERROR [gobuild-minimal 3/3] RUN apt-get update && apt-get install --no-install-recommends -y libseccomp-dev file                                    0.5s
------
 > [git 1/1] RUN apk add --no-cache git xz:
#9 0.707 container_linux.go:370: starting container process caused: exec: "/bin/sh": stat /bin/sh: no such file or directory
------
------
 > [gobuild-minimal 3/3] RUN apt-get update && apt-get install --no-install-recommends -y libseccomp-dev file:
#16 0.437 container_linux.go:370: starting container process caused: exec: "/bin/sh": stat /bin/sh: no such file or directory
------
executor failed running [/bin/sh -c apt-get update && apt-get install --no-install-recommends -y libseccomp-dev file]: exit code: 1
```

From output you can see `apt-get update` runs without the base image being pulled.

tonistiigi commented:
```
diff --git a/builder/builder-next/adapters/containerimage/pull.go b/builder/builder-next/adapters/containerimage/pull.go
index d0891cb598..62977675bd 100644
--- a/builder/builder-next/adapters/containerimage/pull.go
+++ b/builder/builder-next/adapters/containerimage/pull.go
@@ -93,7 +93,11 @@ func (is *Source) resolveRemote(ctx context.Context, ref string, platform *ocisp
 		dgst digest.Digest
 		dt   []byte
 	}
-	res, err := is.g.Do(ctx, ref, func(ctx context.Context) (interface{}, error) {
+	p := platforms.DefaultSpec()
+	if platform != nil {
+		p = *platform
+	}
+	res, err := is.g.Do(ctx, "getconfig::"+ref+"::"+platforms.Format(p), func(ctx context.Context) (interface{}, error) {
 		res := resolver.DefaultPool.GetResolver(is.RegistryHosts, ref, "pull", sm, g)
 		dgst, dt, err := imageutil.Config(ctx, ref, res, is.ContentStore, nil, platform)
 		if err != nil {
@@ -247,7 +251,7 @@ func (p *puller) resolveLocal() {
 }
 
 func (p *puller) resolve(ctx context.Context, g session.Group) error {
-	_, err := p.is.g.Do(ctx, "", func(ctx context.Context) (_ interface{}, err error) {
+	_, err := p.is.g.Do(ctx, "resolve::"+p.ref+"::"+platforms.Format(p.platform), func(ctx context.Context) (_ interface{}, err error) {
 		resolveProgressDone := oneOffProgress(ctx, "resolve "+p.src.Reference.String())
 		defer func() {
 			resolveProgressDone(err)

```

Commit:
builder-next: fix platform-specific behavior

Signed-off-by: Tibor Vass <tibor@docker.com>

tiborvass commented:
@tonistiigi updated

tiborvass commented:
I don't see how TestInspect failure would be related to this PR. Probably a flakey test.

cpuguy83 commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -5608,30 +5608,6 @@ func (s *DockerSuite) TestBuildWithExtraHostInvalidFormat(c *testing.T) {
 
 }
 
-func (s *DockerSuite) TestBuildContChar(c *testing.T) {
```
I don't like that we are removing the test.
Even if it is tested in buildkit, this is more of an end-to-end test, and we make sure it doesn't break regardless of builder.

tiborvass replied:
@cpuguy83 who writes a `\` as the last character of a Dockerfile? I am of the opinion that it's such an edge case that I'm prioritizing removing legacy tests that are useless. But if you feel strongly about this I can fix the test.

cpuguy83 commented on builder/builder-next/adapters/containerimage/pull.go:
```diff
@@ -93,7 +93,11 @@ func (is *Source) resolveRemote(ctx context.Context, ref string, platform *ocisp
+	if platform != nil {
+		p = *platform
+	}
+	res, err := is.g.Do(ctx, "getconfig::"+ref+"::"+platforms.Format(p), func(ctx context.Context) (interface{}, error) {
```
I have no idea what this is doing, likewise in `resolve`. What is this format? What affect does it have?

tiborvass replied:
> Do executes a context function syncronized by the key

https://godoc.org/github.com/moby/buildkit/util/flightcontrol#Group.Do

If the key string is not per-platform, the callback running for a specific platform prevents the run for other platforms when requested simultaneously, that's what I understand at least.

cpuguy83 replied:
Can we make a comment for this, or a function like `func platformKey(op, ref string, platform platform.Platform)`

cpuguy83 commented on builder/builder-next/adapters/containerimage/pull.go:
```diff
@@ -82,10 +83,11 @@ func (is *Source) getResolver(hosts docker.RegistryHosts, ref string, sm *sessio
-	r := resolver.New(hosts, auth)
-	r = is.resolverCache.Add(ref, auth, r, g)
-	return r
+	//auth := resolver.NewSessionAuthenticator(sm, g)
```
Remove commented code

tiborvass replied:
Do you want me to squash https://github.com/moby/moby/commit/3756668adbbb48cf027a2229ba560d3777ed871e with vendor commit? I tried to make that commit understandable to see what it was doing before, while still keeping the vendor commit compilable.

cpuguy83 replied:
Oh I see. I didn't look at the final view, just per commit.

cpuguy83 commented on builder/builder-next/adapters/containerimage/pull.go:
```diff
@@ -67,7 +68,7 @@ type Source struct {
 	is := &Source{
 		SourceOpt:     opt,
-		resolverCache: newResolverCache(),
+		resolverCache: nil, //newResolverCache(),
```
Remove commented code

tiborvass replied:
It is removed in further commits

thaJeztah reviewed (approved):
LGTM

cpuguy83 reviewed (changes requested):
Is there a summary of changes in this bump?

cpuguy83 reviewed (approved):
LGTM
